core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -674,135 +675,232 @@ private RexNode simplifyCoalesce(RexCall call) {
+
+    // delete all branches after the first AlwaysTrue
+    for (int i = 0; i < branches.size(); i++) {
+      CaseBranch branch = branches.get(i);
+      if (branch.cond.isAlwaysTrue()) {
+        branches.subList(i + 1, branches.size()).clear();
+        break;
       }
-      newOperands.add(operand);
     }
-    assert newOperands.size() % 2 == 1;
-    if (newOperands.size() == 1 || values.size() == 1) {
-      final RexNode last = Util.last(newOperands);
-      if (!call.getType().equals(last.getType())) {
-        return rexBuilder.makeAbstractCast(call.getType(), last);
+
+    // collect cardinality of values
+    Set<String> values = branches.stream().map(branch -> {
+      if (unknownAs == FALSE && RexUtil.isNull(branch.value)) {
+        return rexBuilder.makeLiteral(false).toString();
+      } else {
+        return branch.value.toString();
+      }
+    }).collect(Collectors.toSet());
+
+    if (values.size() == 1) {
+      final RexNode firstValue = branches.get(0).value;
+      if (!call.getType().equals(firstValue.getType())) {
+        return rexBuilder.makeAbstractCast(call.getType(), firstValue);
       }
-      return last;
+      return firstValue;
     }
-  trueFalse:
+
     if (call.getType().getSqlTypeName() == SqlTypeName.BOOLEAN) {
-      // Optimize CASE where every branch returns constant true or constant
-      // false.
-      final List<Pair<RexNode, RexNode>> pairs =
-          casePairs(rexBuilder, newOperands);
-      // 1) Possible simplification if unknown is treated as false:
-      //   CASE
-      //   WHEN p1 THEN TRUE
-      //   WHEN p2 THEN TRUE
-      //   ELSE FALSE
-      //   END
-      // can be rewritten to: (p1 or p2)
-      if (unknownAs == FALSE) {
-        final List<RexNode> terms = new ArrayList<>();
-        int pos = 0;
-        for (; pos < pairs.size(); pos++) {
-          // True block
-          Pair<RexNode, RexNode> pair = pairs.get(pos);
-          if (!pair.getValue().isAlwaysTrue()) {
-            break;
-          }
-          terms.add(pair.getKey());
-        }
-        for (; pos < pairs.size(); pos++) {
-          // False block
-          Pair<RexNode, RexNode> pair = pairs.get(pos);
-          if (!pair.getValue().isAlwaysFalse()
-              && !RexUtil.isNull(pair.getValue())) {
-            break;
-          }
-        }
-        if (pos == pairs.size()) {
-          final RexNode disjunction =
-              RexUtil.composeDisjunction(rexBuilder, terms);
-          if (!call.getType().equals(disjunction.getType())) {
-            return rexBuilder.makeCast(call.getType(), disjunction);
-          }
-          return disjunction;
+      final RexNode result = simplifyBooleanCase(rexBuilder, branches, unknownAs, branchType);
+      if (result != null) {
+        if (!call.getType().equals(result.getType())) {
+          return simplify(rexBuilder.makeCast(call.getType(), result), unknownAs);
```

Comment:
Shouldn't we simplify within the cast to make sure we simplify but we preserve the type in the outer expression?

Reply:
I think this might need some cleanup I think this was added around CALCITE-1289 - but after some consideration I've choosen a similar path to what CALCITE-2041 suggested; so the case is only inserted if the call's type was nonNullable but the result is nullable.
I've also moved the simplify call to run on the result instead on the cast.